### PR TITLE
Bugfixes, remove dependency pin

### PR DESCRIPTION
**Pull request by chrisaffirm**
Bugfixes:

- SSH github URLs such as `git@github.com:ReviewNB/jupyterlab-gitplus` did not work. Now they do.
- Some users of reviewnb have custom URLs, often accessible only behind a VPN. For example mine is `https://reviewnb.team.affirm.com`. Now you can configure such URLs.
- This plugin actually works just fine with `jupyter-server>2`, I tested it with `jupyter-server==2.14.0`. I've bumped the pin to jupyter-server<3.

Happy to modify or split this into pieces, I did my best to match existing code style but I do see this is an unsolicited PR.

### CODE REVIEW

**Comment by amit1rrr** [comment]
Thanks @chrisaffirm 

Appreciate taking the time out for this contribution. I'll get to the actual review soon. 

Quick question - which is the latest version of JupyterLab have you tested this against?

**Comment by chrisaffirm** [comment]
>Quick question - which is the latest version of JupyterLab have you tested this against?

These are the major jupyter related libraries I'm using:

```
jupyter-core==5.7.2
jupyter-events==0.10.0
jupyterlab==4.1.6
jupyterhub==4.1.5
jupyter-server[test]==2.14.0
notebook[test]==7.1.3
```

File: jupyterlab_gitplus/utils.py
Comment (amit1rrr) [inline jupyterlab_gitplus/utils.py:6]: Let's just call it `logger` (instead of `_logger`) to be consistent with other log statements in the repo.

File: jupyterlab_gitplus/handlers.py
Comment (amit1rrr) [inline jupyterlab_gitplus/handlers.py:27]: Is this required now?

Since we're not doing anything here wouldn't the base `initialize` method (from  `IPythonHandler`) be enough?

File: jupyterlab_gitplus/__init__.py
Comment (amit1rrr) [inline jupyterlab_gitplus/__init__.py:47]: The only thing that differs between various ReviewNB installations is the actual endpoint (e.g. `https://app.reviewnb.com/` for managed cloud vs `https://reviewnb.team.affirm.com/` for self-hosted. The rest of the URL format / template always remains the same. 

In light of the above we should just call this an optional config as below -

reviewnb_endpoint = nb_server_app.config.get('GitPlus', {}).get('self_hosted_reviewnb_endpoint', 'https://app.reviewnb.com/')

After this - we can just use this `reviewnb_endpoint` config everywhere else in handlers or utils etc. This would minimise code changes since we'd just be replacing hard coded REVIEWNB_ENDPOINT value with this config value.

File: jupyterlab_gitplus/github_v3.py
Comment (amit1rrr) [inline jupyterlab_gitplus/github_v3.py:31]: (after following my top comment above) this would just become - 

```
'reviewnb_url': content['html_url'].replace('https://github.com/', reviewnb_endpoint)
```

File: jupyterlab_gitplus/handlers.py
Comment (amit1rrr) [inline jupyterlab_gitplus/handlers.py:145]: (after following my top comment above) we can just use `reviewnb_endpoint`  in place of `REVIEWNB_ENDPOINT` here.

**Review by amit1rrr — CHANGES_REQUESTED** [review]

**Comment by amit1rrr** [comment]
A lot of our users also have a self-hosted GitHub Enterprise. So bonus points if you can make GitHub endpoint also configurable in the same way as reviewnb endpoint. 

But that's not a pre-requisite for this change. We'd still merge this PR even if only the reviewnb endpoint is configurable. 

Thanks again for your contribution.

File: jupyterlab_gitplus/__init__.py
Comment (amit1rrr) [reply jupyterlab_gitplus/__init__.py:47]: We'd need some validation / check to make sure it's a proper URL. And we probably need to add a trailing slash if not provided e.g. 

some users would set it as - 

```python
c.GitPlus.self_hosted_reviewnb_endpoint = "https://reviewnb.team.affirm.com"
```

& others might set it as -

```python
c.GitPlus.self_hosted_reviewnb_endpoint = "https://reviewnb.team.affirm.com/"
```

File: jupyterlab_gitplus/handlers.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/handlers.py:27]: No, because the base method doesn't take `**kwargs` and we pass them in.

File: jupyterlab_gitplus/utils.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/utils.py]: Done

File: jupyterlab_gitplus/__init__.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/__init__.py:50]: Made this change

File: jupyterlab_gitplus/handlers.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/handlers.py:145]: Done

**Comment by chrisaffirm** [comment]
Hi @amit1rrr please take another look, I think I've addressed all your concerns. 

Happy to make the change for github endpoints, but one point I'm not sure of (I don't have a self-hosted github so can't check). You have, as constants in `handlers.py`, both GITHUB_ENDPOINT and GITHUB_REMOTE_DOMAIN.

Are these always synchronized, i.e. `GITHUB_ENDPOINT=f"https://{GITHUB_REMOTE_DOMAIN}/"`? If so it seems like a straightforward change.

File: jupyterlab_gitplus/handlers.py
Comment (amit1rrr) [inline jupyterlab_gitplus/handlers.py:166]: it'd be consistent to simply accept the "context" dict & fetch the required values from it (like below). Right now  we're passing the same values to each handler but accepting those in each handler slightly differently.

```
class ServerConfigHandler(IPythonHandler):
    def initialize(self, context):
        self.server_root_dir = context['server_root_dir']
```

```
class CommitHandler(IPythonHandler):
    def initialize(self, context):
        self.github_token = context['github_token']
        self.reviewnb_endpoint = context['reviewnb_endpoint']
```        
        
 ```
 class PullRequestHandler(IPythonHandler):
    def initialize(self, context):
        self.github_token = context['github_token']
        self.reviewnb_endpoint = context['reviewnb_endpoint']
```

File: jupyterlab_gitplus/github_v3.py
Comment (amit1rrr) [inline jupyterlab_gitplus/github_v3.py:31]: missing a trailing slash in github.com which is creating 2 slashes after the endpoint as shown in screenshot. The line should be 

```
"reviewnb_url": content['html_url'].replace('https://github.com/', reviewnb_endpoint)
```

Screenshot 2024-06-10 at 7 04 15 PM

File: jupyterlab_gitplus/__init__.py
Comment (amit1rrr) [inline jupyterlab_gitplus/__init__.py:57]: It'd be worth adding a logger.info statement indicating the `reviewnb_endpoint` value set in the current set up.

File: README.md
Comment (amit1rrr) [inline README.md:51]: Let's please call this `c.GitPlus.self_hosted_reviewnb_endpoint` to be explicit.

File: README.md
Comment (amit1rrr) [inline README.md:45]: #### 3. [OPTIONAL] Configuration for Self Hosted ReviewNB


##### 3.1 Configuring Self Hosted ReviewNB URL
.......


(in the future)
##### 3.2 Configuring Self Hosted GitHub Enterprise Server URL
.......

**Comment by amit1rrr** [comment]
> Hi @amit1rrr please take another look, I think I've addressed all your concerns.
> 
> Happy to make the change for github endpoints, but one point I'm not sure of (I don't have a self-hosted github so can't check). You have, as constants in `handlers.py`, both GITHUB_ENDPOINT and GITHUB_REMOTE_DOMAIN.
> 
> Are these always synchronized, i.e. `GITHUB_ENDPOINT=f"https://{GITHUB_REMOTE_DOMAIN}/"`? If so it seems like a straightforward change.


Yes, they're always the same. There's also a bit of change for `GITHUB_REST_ENDPOINT` in `github_v3.py` which should become -

```
GITHUB_REST_ENDPOINT = urllib.parse.urljoin(get_env_variable('GITHUB_BASE_URL'), 'api/v3')
```

I'd be inclined to skip this GitHub URL change for another time as it might lead us to more rounds of reviews & testing before we release the main change you need.

File: jupyterlab_gitplus/__init__.py
Comment (amit1rrr) [inline jupyterlab_gitplus/__init__.py:8]: My local build kept failing until I removed this statement. 

Not related to your change but it seems like the original state was broken i.e. it's referring to `._version` which does not exist & is not even required since we explicitly set the version on line 31 below this.

File: jupyterlab_gitplus/handlers.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/handlers.py]: I've updated the PR to do it this way, but I'm not sure it's cleaner. 

Note how in the init we need to do `{"context": context}` because the last arg of `web_app.add_handlers` is expected to be kwargs.

File: jupyterlab_gitplus/github_v3.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/github_v3.py:31]: I've made this fix.

Also, thinking about it, for a user with a custom github domain is likely to have`content["html_url"]` start with `https://github.yourcompany.com" instead of "github.com". Or something like that.

Since I can't test any of this I'll definitely hold off adding the github custom domain functionality.

File: jupyterlab_gitplus/__init__.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/__init__.py:57]: Done

File: README.md
Comment (chrisaffirm) [reply README.md]: Done

File: jupyterlab_gitplus/__init__.py
Comment (chrisaffirm) [reply jupyterlab_gitplus/__init__.py:8]: In my *installation* of jupyterlab gitplus, `_version.py` exists and has these contents:

```
import json
from pathlib import Path

__all__ = ["__version__"]

def _fetchVersion():
    HERE = Path(__file__).parent.resolve()

    for settings in HERE.rglob("package.json"): 
        try:
            with settings.open() as f:
                return json.load(f)["version"]
        except FileNotFoundError:
            pass

    raise FileNotFoundError(f"Could not find package.json under dir {HERE!s}")

__version__ = _fetchVersion()
```

The actual version info is in `jupyterlab_gitplus/labextension/package.json`. So I'm guessing you are trying to run the code directly from the git repo, while this stuff gets added in by the jupyterlab extension packaging system. 

I'll leave it alone for now, but can remove it if you want to override ⬆️  behavior (or if my guess as to what's happening is wrong).

File: README.md
Comment (chrisaffirm) [reply README.md:45]: Done

**Comment by chrisaffirm** [comment]
I think I've addressed all your change requests, apart from the the `_version.py` one.

**Comment by chrisaffirm** [comment]
Bump on this? Would love to merge and deploy this (instead of internal fork) to my prod.

**Comment by amit1rrr** [comment]
@chrisaffirm apologies for the delay. I've just merged the PR. Will create a release soon.

**Comment by chrisaffirm** [comment]
Thanks so much! Looking forward to getting back on main.

**Comment by amit1rrr** [comment]
@chrisaffirm We've released the latest version for Jupyterlab GitPlus which has your changes in it. You can view/install the new version here - https://pypi.org/project/jupyterlab-gitplus/0.3.4/

Thank you for your contribution. 

Please feel free to reach out here or on email (amit at reviewnb dt com) if you face any issues.

**Pull request merged**
